File: sdk/go/signature.go
package relay

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strconv"
	"time"
)

const (
	// SignatureHeader is the header containing the webhook signature.
	SignatureHeader = "X-Relay-Signature"
	// TimestampHeader is the header containing the request timestamp.
	TimestampHeader = "X-Relay-Timestamp"
	// DefaultTimestampTolerance is the default tolerance for timestamp validation.
	DefaultTimestampTolerance = 5 * time.Minute
)

// VerifySignature verifies a webhook signature.
// payload is the raw request body, signature is from X-Relay-Signature header,
// timestamp is from X-Relay-Timestamp header, and secret is your signing secret.
func VerifySignature(payload []byte, signature, timestamp, secret string) error {
	return VerifySignatureWithTolerance(payload, signature, timestamp, secret, DefaultTimestampTolerance)
}

// VerifySignatureWithTolerance verifies a webhook signature with custom timestamp tolerance.
func VerifySignatureWithTolerance(payload []byte, signature, timestamp, secret string, tolerance time.Duration) error {
	// Validate timestamp
	ts, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return fmt.Errorf("invalid timestamp format: %w", err)
	}

	reqTime := time.Unix(ts, 0)
	age := time.Since(reqTime)
	if age < -tolerance || age > tolerance {
		return fmt.Errorf("timestamp outside acceptable range: %v old", age)
	}

	// Compute expected signature
	expected := ComputeSignature(payload, timestamp, secret)

	// Compare signatures using constant-time comparison
	if !hmac.Equal([]byte(signature), []byte(expected)) {
		return fmt.Errorf("signature mismatch")
	}

	return nil
}

// VerifySignatureWithSecrets verifies a webhook signature against multiple secrets.
// Useful during secret rotation when both old and new secrets are valid.
func VerifySignatureWithSecrets(payload []byte, signature, timestamp string, secrets []string, tolerance time.Duration) error {
	for _, secret := range secrets {
		err := VerifySignatureWithTolerance(payload, signature, timestamp, secret, tolerance)
		if err == nil {
			return nil
		}
	}
	return fmt.Errorf("signature does not match any provided secret")
}

// ComputeSignature computes the HMAC-SHA256 signature for a payload.
func ComputeSignature(payload []byte, timestamp, secret string) string {
	// Signature is computed over: timestamp + "." + payload
	message := timestamp + "." + string(payload)

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(message))
	return hex.EncodeToString(mac.Sum(nil))
}

// WebhookVerifier provides a convenient way to verify webhooks with a fixed secret.
type WebhookVerifier struct {
	secrets   []string
	tolerance time.Duration
}

// NewWebhookVerifier creates a new webhook verifier with the given secret.
func NewWebhookVerifier(secret string) *WebhookVerifier {
	return &WebhookVerifier{
		secrets:   []string{secret},
		tolerance: DefaultTimestampTolerance,
	}
}

// NewWebhookVerifierWithSecrets creates a verifier that accepts multiple secrets.
// Useful during secret rotation.
func NewWebhookVerifierWithSecrets(secrets []string) *WebhookVerifier {
	return &WebhookVerifier{
		secrets:   secrets,
		tolerance: DefaultTimestampTolerance,
	}
}

// WithTolerance sets the timestamp tolerance.
func (v *WebhookVerifier) WithTolerance(d time.Duration) *WebhookVerifier {
	v.tolerance = d
	return v
}

// Verify verifies a webhook request.
func (v *WebhookVerifier) Verify(payload []byte, signature, timestamp string) error {
	return VerifySignatureWithSecrets(payload, signature, timestamp, v.secrets, v.tolerance)
}

File: sdk/typescript/src/signature.ts
import { createHmac, timingSafeEqual } from 'crypto';
import {
  InvalidSignatureError,
  InvalidTimestampError,
  MissingSignatureError,
  MissingTimestampError,
} from './errors';

const DEFAULT_TOLERANCE_SECONDS = 300; // 5 minutes

export interface VerifyOptions {
  toleranceSeconds?: number;
  currentTimestamp?: number;
}

export interface WebhookHeaders {
  signature: string;
  timestamp: string;
}

/**
 * Verify a webhook signature from Relay.
 *
 * @param payload - The raw request body as a string or Buffer
 * @param signature - The X-Relay-Signature header value
 * @param timestamp - The X-Relay-Timestamp header value
 * @param secret - Your webhook signing secret
 * @param options - Optional verification options
 * @throws {MissingSignatureError} If signature is missing
 * @throws {MissingTimestampError} If timestamp is missing
 * @throws {InvalidTimestampError} If timestamp is invalid or too old
 * @throws {InvalidSignatureError} If signature doesn't match
 */
export function verifySignature(
  payload: string | Buffer,
  signature: string,
  timestamp: string,
  secret: string,
  options: VerifyOptions = {}
): void {
  if (!signature) {
    throw new MissingSignatureError();
  }
  if (!timestamp) {
    throw new MissingTimestampError();
  }

  const tolerance = options.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS;
  const now = options.currentTimestamp ?? Math.floor(Date.now() / 1000);

  // Parse timestamp
  const ts = parseInt(timestamp, 10);
  if (isNaN(ts)) {
    throw new InvalidTimestampError('Timestamp is not a valid number');
  }

  // Check timestamp is within tolerance
  if (Math.abs(now - ts) > tolerance) {
    throw new InvalidTimestampError(
      `Timestamp is outside the tolerance window (${tolerance}s)`
    );
  }

  // Compute expected signature
  const payloadStr = typeof payload === 'string' ? payload : payload.toString('utf8');
  const signedPayload = `${timestamp}.${payloadStr}`;
  const expectedSignature = computeSignature(signedPayload, secret);

  // Compare signatures using timing-safe comparison
  if (!secureCompare(signature, expectedSignature)) {
    throw new InvalidSignatureError();
  }
}

/**
 * Verify a webhook using multiple secrets (for secret rotation).
 * Returns true if ANY secret validates the signature.
 *
 * @param payload - The raw request body
 * @param signature - The X-Relay-Signature header value
 * @param timestamp - The X-Relay-Timestamp header value
 * @param secrets - Array of secrets to try (current + previous)
 * @param options - Optional verification options
 * @throws {MissingSignatureError} If signature is missing
 * @throws {MissingTimestampError} If timestamp is missing
 * @throws {InvalidTimestampError} If timestamp is invalid or too old
 * @throws {InvalidSignatureError} If no secret validates the signature
 */
export function verifySignatureWithRotation(
  payload: string | Buffer,
  signature: string,
  timestamp: string,
  secrets: string[],
  options: VerifyOptions = {}
): void {
  if (!signature) {
    throw new MissingSignatureError();
  }
  if (!timestamp) {
    throw new MissingTimestampError();
  }
  if (!secrets.length) {
    throw new InvalidSignatureError();
  }

  const tolerance = options.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS;
  const now = options.currentTimestamp ?? Math.floor(Date.now() / 1000);

  // Parse and validate timestamp
  const ts = parseInt(timestamp, 10);
  if (isNaN(ts)) {
    throw new InvalidTimestampError('Timestamp is not a valid number');
  }
  if (Math.abs(now - ts) > tolerance) {
    throw new InvalidTimestampError(
      `Timestamp is outside the tolerance window (${tolerance}s)`
    );
  }

  // Try each secret
  const payloadStr = typeof payload === 'string' ? payload : payload.toString('utf8');
  const signedPayload = `${timestamp}.${payloadStr}`;

  for (const secret of secrets) {
    const expectedSignature = computeSignature(signedPayload, secret);
    if (secureCompare(signature, expectedSignature)) {
      return; // Valid signature found
    }
  }

  throw new InvalidSignatureError();
}

/**
 * Compute an HMAC-SHA256 signature in the Relay format.
 *
 * @param payload - The signed payload (timestamp.body)
 * @param secret - The signing secret
 * @returns The signature in "v1=<hex>" format
 */
export function computeSignature(payload: string, secret: string): string {
  const hmac = createHmac('sha256', secret);
  hmac.update(payload);
  return `v1=${hmac.digest('hex')}`;
}

/**
 * Extract webhook headers from a request headers object.
 *
 * @param headers - Headers object (works with various frameworks)
 * @returns The signature and timestamp headers
 */
export function extractWebhookHeaders(
  headers: Record<string, string | string[] | undefined> | Headers
): WebhookHeaders {
  const getHeader = (name: string): string => {
    if (headers instanceof Headers) {
      return headers.get(name) ?? '';
    }
    const value = headers[name] ?? headers[name.toLowerCase()];
    return Array.isArray(value) ? value[0] : value ?? '';
  };

  return {
    signature: getHeader('X-Relay-Signature'),
    timestamp: getHeader('X-Relay-Timestamp'),
  };
}

// Timing-safe string comparison
function secureCompare(a: string, b: string): boolean {
  try {
    const bufA = Buffer.from(a);
    const bufB = Buffer.from(b);
    if (bufA.length !== bufB.length) {
      return false;
    }
    return timingSafeEqual(bufA, bufB);
  } catch {
    return false;
  }
}
